pkg/controller/installlogmonitor/installlogmonitor_controller.go

```diff
@@ -0,0 +1,280 @@
+	}
+
+	iLog.Info("processing new install log")
+
+	// Apply the processed annotation and save before we report anything. If we succeed here and fail later
+	// we accept that we will lose processing of that install log. This is for tracking/metrics and not
+	// anything critical.
+	if cm.ObjectMeta.Annotations == nil {
+		cm.ObjectMeta.Annotations = map[string]string{}
+	}
+	cm.ObjectMeta.Annotations[processedAnnotation] = "true"
+	if err := r.Client.Update(context.Background(), cm); err != nil {
+		iLog.WithError(err).Error("error saving processed annotation")
+		return reconcile.Result{}, err
+	}
+
+	success, successKeyReported := cm.Data["success"]
+
+	foundError := false
+	if !successKeyReported || success == "false" {
+		log := cm.Data["log"]
+		// Log each line separately, this brings all our install logs from many namespaces into
+		// the main hive log where we can aggregate search results.
+		installLogLines := strings.Split(log, "\n")
+		for _, l := range installLogLines {
+			iLog.WithField("line", l).Info("install log line")
+		}
+
+		// Scan log contents for known errors if the install was not reported as a success:
+		for _, ilr := range ilRegexes {
+			for _, re := range ilr.SearchRegexes {
+				if re.Match([]byte(log)) {
+					iLog.WithField("reason", ilr.InstallFailingReason).Info("found known install failure string")
+					cd.Status.Conditions = controllerutils.SetClusterDeploymentCondition(cd.Status.Conditions, hivev1.InstallFailingCondition,
+						corev1.ConditionTrue, ilr.InstallFailingReason, ilr.InstallFailingMessage, controllerutils.UpdateConditionAlways)
+					// Increment a counter metric for this cluster type and error reason:
+					metricInstallErrors.WithLabelValues(hivemetrics.GetClusterDeploymentType(cd), ilr.InstallFailingReason).Inc()
+					foundError = true
+					break
+				}
+			}
+			if foundError {
+				break
+			}
+		}
+	}
+
+	// If we reach this point, we have found no errors. If the install log configmap specifies that the install failed,
+	// we will still report InstallFailed condition with reason unknown. Note that older install manager pods will not
+	// set success, so for these we will just assume success and no condition will be set.
+	if !foundError && successKeyReported && success == "false" {
+		iLog.Info("install failed but no errors strings found, reporting unknown error")
+		cd.Status.Conditions = controllerutils.SetClusterDeploymentCondition(cd.Status.Conditions, hivev1.InstallFailingCondition,
+			corev1.ConditionTrue, unknownReason, unknownMessage, controllerutils.UpdateConditionAlways)
+		metricInstallErrors.WithLabelValues(hivemetrics.GetClusterDeploymentType(cd), unknownReason).Inc()
+	} else if !foundError {
+		// Assume success and clear the condition if it is present.
+		iLog.Info("install successful, clearing InstallFailing condition if set")
+		cd.Status.Conditions = controllerutils.SetClusterDeploymentCondition(cd.Status.Conditions, hivev1.InstallFailingCondition,
```

Comment:
what if you got an error, it's a legacy configmap (so no success key), and the error doesn't match any of your regexes? 
Should we only clear the error condition if .Status.Installed == true ?

Reply:
I thought about this but there's a delay issue there, the install log configmap gets uploaded by the install pod itself, the Installed bool gets set in clusterdeployment controller when the job ends, so the ConfigMap is very likely processed first when Installed is still false. Implication/tradeoff here would be that installs using old images will not get Unknown failure conditions.

Reply:
Ack, makes sense. If we set an error condition, we'll already be using the new code which will set a success key.